Src/Support/Google.Apis.Auth/SignedTokenVerification.cs

```diff
@@ -0,0 +1,249 @@
+            {
+                certificatesLocation.ThrowIfNullOrEmpty(nameof(certificatesLocation));
+                certificateFactory.ThrowIfNull(nameof(certificateFactory));
+
+                await _cacheLock.WaitAsync(cancellationToken).ConfigureAwait(false);
+                try
+                {
+                    if (forceCertificateRefresh ||
+                        !_cache.TryGetValue(certificatesLocation, out CachedCertificates cachedCerts) ||
+                        cachedCerts.Expired(_clock.UtcNow))
+                    {
+                        string certificatesJson = await FetchCertificatesAsync(certificatesLocation).ConfigureAwait(false);
+                        cachedCerts = new CachedCertificates(
+                            JToken.Parse(certificatesJson)["keys"]
+                            .AsEnumerable()
+                            .Select(key => certificateFactory(key)).ToList(),
+                            _clock.UtcNow);
+                        _cache[certificatesLocation] = cachedCerts;
+                    }
+                    return cachedCerts.Certificates;
+                }
+                finally
+                {
+                    _cacheLock.Release();
+                }
+            }
+
+            // Abstract for testing purposes.
+            protected abstract Task<string> FetchCertificatesAsync(string certificatesLocation);
+
+            internal struct CachedCertificates
+            {
+                private readonly DateTimeOffset _cachedSince;
+                public IEnumerable<object> Certificates { get; }
+
+                public CachedCertificates(IEnumerable<object> certificates, DateTimeOffset cachedSince)
+                {
+                    _cachedSince = cachedSince;
+                    Certificates = certificates;
+                }
+
+                
+                public bool Expired(DateTimeOffset now) => _cachedSince + DefaultCacheValidity <= now;
+            }
+
+            #region IDisposable Support
+            private bool disposedValue = false; // To detect redundant calls
+
+            void Dispose(bool disposing)
+            {
+                if (!disposedValue)
+                {
+                    if (disposing)
+                    {
+                        _cacheLock.Dispose();
+                    }
+
+                    _cache.Clear();
+                    disposedValue = true;
+                }
+            }
+
+            public void Dispose() => Dispose(true);
+            #endregion
+        }
+
+        internal sealed class CertificateCache : CertificateCacheBase
+        {
+            public CertificateCache()
+                :base(SystemClock.Default)
+            { }
+
+            protected override async Task<string> FetchCertificatesAsync(string certificatesLocation)
+            {
+                string certificatesJson;
+                using (var httpClient = new HttpClient())
```

Comment:
No need to have a local variable here - just return in the `using` statement.

This could also be done without a separate block now:

```csharp
using var httpClient = new HttpClient();
return await httpClient.GetStringAsync(certificatesLocation).ConfigureAwait(false);
```

As an aside, do we definitely *want* to create a new HttpClient here? I'm thinking about proxies etc...